### ISSUE TITLE
incorrect usage of detach/cpu/to

### ISSUE THREAD

**Issue by dvolgyes** [PyTorchLightning/pytorch-lightning#6214]
## 🐛 Bug

Incorrect use of detach() and cpu() during fixing #4592.

## Please reproduce using the BoringModel

You cannot really.

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

The fix for #4592 has good intentions but an obvious bug slipped through.

It is easy to understand but hard to test it, so let's rely on common sense:
That is the original code:
```
            if isinstance(output, Result):
                output.detach()
                if self.move_metrics_to_cpu:
                    output.cpu()
```

### Expected behavior

And that supposed to be the correct code, obviously:
```
            if isinstance(output, Result):
                output = output.detach()
                if self.move_metrics_to_cpu:
                    output = output.cpu()
```
### Environment
Any.

### Additional context

I will make a trivial PR for this, I just wanted to create a ticket to actually track the issue.
In theory, the bug is general pattern, but in practice these two lines were the only ones affected in PL.

**Comment by dvolgyes** [comment]
I think the proposed fix is obvious, review is wanted, and after approval, it could be merged, I have nothing further for this ticket.

**Comment by dvolgyes** [comment]
I actually found a few other occasions too, see followup commits.

**Comment by Borda** [comment]
@dvolgyes great, mind sending a PR? 🐰

**Comment by dvolgyes** [comment]
I see that some tests fail, but I don't see how is this related to my changes.
Can you help me understand at least one of the test cases? 
Or is it just timeout?

**Comment by edenlightning** [comment]
@Borda

**Comment by tchaton** [comment]
Dear @dvolgyes,

Result objects were doing an in-place operation. Reverted your changes on the model as it is cleaner to use them as in-place operation. 
However, Result objects were just dictionaries, but it makes sense to have return self.

Best,
T.C
